### ISSUE TITLE
Declaring the type of `m` in a function

### ISSUE THREAD

**Issue by markwpearce** [rokucommunity/brighterscript#1036]
In order to give type hints for `m` in component functions, callbacks, etc, it will be useful to be able to specify/declare its type

the proposal for this currently is to add `type m as _______` after the function signature....:

eg, for this  interface declaration:
```brighterscript
interface MyComponentObject
   top as roSGNodeMyComponent
   isEnabled as boolean
   isSelected as boolean
end interface
```

you can specify `m` as that type in this way:

```brighterscript
sub printDetails() type m as MyComponentObject
   print m.top.text, m.isEnabled
end sub
```

```brighterscript
function handleButtonPress() as boolean, type m as MyComponentObject
   ' we know m is MyComponentObject, m.top is roSGNodeMyComponent, etc.
   if m.isEnabled 
       m.isSelected = true
       m.top.text = "Clicked"
   end if
   return m.isEnabled
end function
```

```brighterscript
sub callOtherFunc()
   doSomethingElse(function() type m as MyComponentObject
           ' this is a call back
           print m.top.text
       end function, otherParam)
end sub
```

**Comment by TwitchBronBron** [comment]
Here's the latest proposal. I think we should move forward with this.

Syntax:
```brighterscript
type m as SomeInterface
```
**Things to consider:**
 - class methods (and functions on AALiterals) retain their own `m` type. But anon functions inside of them would inherit the closest m type override
 - the variable being type-overridden must already exist (i.e. we are not declaring _new_ types)
 - only runtime variables can be type overridden. (i.e. can't do this for classes, interfaces, enums, etc). This basically means it only really applies to `m`, but if closures every show up.
 - These statements _must_ be declared at the _top_ of their enclosing scope (top of file, top of namespace, top of function). They can be intermixed with import statements at top of file.

**Comment by markwpearce** [comment]
fixed with #1140
